Jasmine/Grid/ViewControllers/GridGameViewController.swift

```diff
@@ -0,0 +1,221 @@
+        gridCollectionView.performBatchUpdates(gridCollectionView.reloadData,
+                                                     completion: nil)
+    }
+
+    /* Gesture Recognisers */
+    /// Listens to a drag gesture and handles the operation of dragging a tile, and dropping it
+    /// to another location.
+    @IBAction func onTilesDragged(_ sender: UIPanGestureRecognizer) {
+        let position = sender.location(in: gridCollectionView)
+
+        switch sender.state {
+        case .began:
+            handleTileSelected(at: position)
+
+        case .changed:
+            handleTileDragged(at: position)
+
+        case .ended:
+            handleTileLanding(at: position)
+
+        default:
+            break
+        }
+    }
+}
+
+// MARK: - Drag and Drop Tiles
+fileprivate extension GridGameViewController {
+
+    /// Handles the case when the tile is lifted from the collection view.
+    /// Such a tile gets "detached" from the view by having its identity noted down in the 
+    /// properties.
+    ///
+    /// - Parameter position: location where the tile is selected.
+    fileprivate func handleTileSelected(at position: CGPoint) {
+        guard draggingTile == nil else {
+            return
+        }
+        guard draggingStartFrame == nil else {
+            return
+        }
+        guard let indexTouched = gridCollectionView.indexPathForItem(at: position) else {
+            return
+        }
+        guard let cellTouched = gridCollectionView
+            .cellForItem(at: indexTouched) as? SquareTextViewCell else {
+                return
+        }
+
+        gridCollectionView.bringSubview(toFront: cellTouched)
+        draggingTile = cellTouched
+        draggingStartFrame = cellTouched.frame
+        draggingStartIndex = indexTouched
+    }
+
+    /// Drags the tile that is referenced in `draggingTile`.
+    ///
+    /// - Parameter position: location where the tile should drag to.
+    fileprivate func handleTileDragged(at position: CGPoint) {
+        guard let draggingTile = draggingTile else {
+            return
+        }
+        let size = draggingTile.frame.size
+        let newOrigin = GeometryUtils.getOrigin(from: position, withSize: size)
+        draggingTile.frame.origin = newOrigin
+    }
+
+    /// Lands the tile at the specified position, if such a cell is found.
+    /// The other cell will snap to the dragged cell position.
+    /// If fails to find a place, the dragged cell shall return to original position.
+    ///
+    /// - Parameter position: location where the tile should land.
+    fileprivate func handleTileLanding(at position: CGPoint) {
+        guard let indexLanded = gridCollectionView.indexPathForItem(at: position) else {
+            handleTileFailedLanding()
+            return
+        }
+        guard let cellToVacate = gridCollectionView
+            .cellForItem(at: indexLanded) as? SquareTextViewCell else {
+                handleTileFailedLanding()
+                return
```

Comment:
still think should indent one more, as it's a different flow path than the above line.